pkg/finalizer/finalizer.go

```diff
@@ -0,0 +1,59 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package finalizer
+
+import (
+	"context"
+	"fmt"
+
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+type finalizers map[string]Finalizer
+
+// NewFinalizers returns the Finalizers interface
+func NewFinalizers() Finalizers {
+	return finalizers{}
+}
+
+func (f finalizers) Register(key string, finalizer Finalizer) error {
+	if _, ok := f[key]; ok {
+		return fmt.Errorf("finalizer for key %q already registered", key)
+	}
+	f[key] = finalizer
+	return nil
+}
+
+func (f finalizers) Finalize(ctx context.Context, obj client.Object) (bool, error) {
+	var errList []error
+	needsUpdate := false
+	for key, finalizer := range f {
+		if dt := obj.GetDeletionTimestamp(); dt.IsZero() && !controllerutil.ContainsFinalizer(obj, key) {
+			controllerutil.AddFinalizer(obj, key)
+			needsUpdate = true
+		} else if !dt.IsZero() && controllerutil.ContainsFinalizer(obj, key) {
+			ret, err := finalizer.Finalize(ctx, obj)
```

Comment:
Nit: more obvious name
```suggestion
			finalizerNeedsUpdate, err := finalizer.Finalize(ctx, obj)
```